### DISCUSSION TITLE
How to disable pylance's definition provider but keep other providers

### DISCUSSION THREAD

**Discussion by c3-lukelewis** [microsoft/pylance-release#3866] [Q&A]
Is there a way to disable pylance's `Go To Definition` provider in favor of a custom one we implemented, but keep the other providers? 

I understand that vscode will merge provider results if there are multiple, but we do not wan't to show the result from pylance's `Go To Definition` provider.

**Comment by rchiodo** [comment]
Sorry but there's no builtin way to disable the definition provider. At least that's built into pylance itself.

I don't _think_ VS code has a way to filter them out either, but you might try asking there.  Although, if they did it would be controlled by the user, not be an extension.

The hacky way would be to run the pyright language server on your own and you provide the goto definition support but then forward everything else to pyright.

**Reply by erictraut** [reply]
The Pyright VS Code extension supports a setting called `pyright.disableLanguageServices` that allows you to disable all language service features. This would allow you to use pyright's type checking capabilities with none of the language server features. There is no way to enable or disable individual language server features.

**Reply by c3-lukelewis** [reply]
Got it. Thanks for the input

**Reply by c3-lukelewis** [reply]
@erictraut would we be able to build our own language service on top of pyright's? pyright's language services provides us almost all of the heavy lifting needed, however we just need to tack on some final redirects/manipulations based on their return values (we would like to redirect to a non-python file where we "define" our types)

It would be perfect if we could use `pyright.disableLanguageServices` while keeping the type checking capabilities as you mentioned prior, but also use our own "language services" above

**Reply by erictraut** [reply]
Can you explain more about what you're trying to do? Why do you feel the need to implement your own "Go To Definition" provider?

**Reply by c3-lukelewis** [reply]
We have our own type definition language that can be implemented in python. We want to be able to redirect to those files. They have their own extensions.

For example `RaceCar.ext`
```
type RaceCar:
   vroom: function()
```
RaceCar's `vroom` can be implemented in python

We generate a RaceCar.pyi file from the `.ext` files, which pyright is able to understand to give us type checking, parsing, etc. 

The Go To's will take us to `RaceCar.pyi`, but what we really want is to redirect to `RaceCar.ext`.

For developer experience we want to avoid going to the `.pyi` files as much as we can, and go straight to `.ext`

**Reply by erictraut** [reply]
Sounds like your needs are very specific. Pyright is open source, so you could build a custom solution using it as long as you don't violate the license terms. You should consider whether maintaining your own custom solution is worth it because it will require ongoing maintenance. We won't be in a position to provide support or bug fixes, especially if your solution falls out of date with the mainline pyright code.

**Reply by c3-lukelewis** [reply]
Thanks, that makes sense.
